### PR TITLE
implemented side by side rendering with y flag

### PR DESCRIPTION
**Pull request by coltonoscopy**
This currently works for small files because it renders to a temporary PDF before merging the first two pages, but for larger files (where the first file will bleed over into the second page), how do we want to handle?

### CODE REVIEW

**Comment by dmalan** [comment]
Most files will be > 1 page. Can we not iterate over each PDF's pages and concatenate pairs? Or let weasyprint render them using HTML frames that are dynamically generated in a temp file?

**Comment by coltonoscopy** [comment]
We can probably do that with a little work; what do you want to have happen when the page #s don't line up? Render blank pages on the shorter PDF's side?

File: render50 [hunk 1]
```diff
@@ -44,6 +45,7 @@ def main():
     parser.add_argument("--no-color", action="store_true", help="disable syntax highlighting")
     parser.add_argument("-s", "--size", help="size of page")
     parser.add_argument("-x", "--exclude", action="append", help="pattern to exclude")
+    parser.add_argument("-y", "--side-by-side", action="store_true", help="render two files side-by-side")
```
Comment (dmalan) [inline render50]: tweaked to be consistent with `diff`

File: render50 [hunk 2]
```diff
@@ -53,12 +55,19 @@ def main():
     if not output.lower().endswith(".pdf"):
         output += ".pdf"
 
+    # Ensure there are two and only two input files if side-by-side specified
+    if args.side_by_side and len(args.input) != 2:
+        raise RuntimeError("Can only render 2 files side-by-side.")
```
Comment (dmalan) [inline render50]: tweaked error message

File: render50 [hunk 3]
```diff
@@ -53,12 +55,19 @@ def main():
...
+
     # Check for size
     size = args.size
     if not size:
         size = "letter landscape"
     elif size == "A4" or size == "letter":
         size = "{} landscape".format(args.size.strip())
+    else:
+        raise RuntimeError("Can only render 2 files side-by-side.")
```
Comment (dmalan) [inline render50]: old bug

File: render50 [hunk 4]
```diff
@@ -198,12 +207,76 @@ def main():
...
     if documents:
-        pages = [page for document in documents for page in document.pages]
-        documents[0].copy(pages).write_pdf(output)
+
+        # If we're in side-by-side mode, write two temp PDFs, then concatenate them
+        if args.side_by_side:
+
+            # Write temp PDFs for first and second document
+            documents[0].write_pdf('temp1.pdf')
+            documents[1].write_pdf('temp2.pdf')
```
Comment (dmalan) [inline render50]: Definitely don't use `$PWD`, use [`tempfile.mkstemp`](https://docs.python.org/2/library/tempfile.html#tempfile.mkstemp).

File: render50 [hunk 5]
```diff
@@ -198,12 +207,76 @@ def main():
...
         cprint("Rendered {}.".format(output), "green")
     else:
         cprint("Nothing to render.", "red")
 
+def merge_pdfs(output):
+    """ 
+    Combines two temp PDFs page by page, side by side, filling in blank pages for shorter PDF side. 
+    https://stackoverflow.com/questions/24580454/python-pypdf2-join-pages/34646677#34646677
+    """
+    with open('temp1.pdf', 'rb') as temp1, open('temp2.pdf', 'rb') as temp2, open('blank.pdf', 'rb') as blank:
```
Comment (dmalan) [inline render50]: Not appropriate to hardcode filenames here, breaks abstraction. Should be taken in as parameters.

Comment (dmalan) [reply render50]: Do use double quotes for consistency.

File: render50 [hunk 6]
```diff
@@ -198,12 +207,76 @@ def main():
...
+        # Load input pdfs
+        input_pdf_1 = PyPDF2.PdfFileReader(temp1)
+        input_pdf_2 = PyPDF2.PdfFileReader(temp2)
+        blank_pdf = PyPDF2.PdfFileReader(blank)
```
Comment (dmalan) [inline render50]: Let's create and delete `blank.pdf` on the fly so that its dimensions match `args.size`.

File: render50 [hunk 7]
```diff
@@ -198,12 +207,76 @@ def main():
...
+
+        # Determine which PDF is larger
+        temp1_pages = input_pdf_1.getNumPages()
+        temp2_pages = input_pdf_2.getNumPages()
+        pages_to_iterate = max(temp1_pages, temp2_pages)
```
Comment (dmalan) [inline render50]: Instead of `temp1` and `temp2` terminology, let's use `left` and `right` for clarity.

File: render50 [hunk 8]
```diff
@@ -53,12 +55,19 @@ def main():
...
+    # Ensure there are two and only two input files if side-by-side specified
+    if args.side_by_side and len(args.input) != 2:
+        raise RuntimeError("Can only render 2 files side-by-side.")
+
     # Check for size
     size = args.size
     if not size:
         size = "letter landscape"
     elif size == "A4" or size == "letter":
         size = "{} landscape".format(args.size.strip())
```
Comment (dmalan) [inline render50]: In side-by-side mode, let's default to `{} portrait`.

File: render50 [hunk 9]
```diff
@@ -198,12 +207,76 @@ def main():
...
+            else:
+                second_pdf = input_pdf_2.getPage(page_num)
+
+            # Dimensions for offset from loaded page (adding it to the right)
+            offset_x = output_pdf.mediaBox[2]
+            offset_y = 0
+
+            # Add second page to first one
+            output_pdf.mergeTranslatedPage(second_pdf, offset_x, offset_y, expand=True)
+            completed_pages.append(output_pdf)
```
Comment (dmalan) [inline render50]: Can we also scale the `output_pdf` to be `{} landscape`, where `{}` is `args.size`? Perhaps with https://pythonhosted.org/PyPDF2/PageObject.html?highlight=mergetranslatedpage#PyPDF2.pdf.PageObject.scaleTo?

I.e., if user specifies `A4`, then we should render each of the left and right in `A4 portrait`, then merge them side-by-side in order to output an `A4 landscape` result.

I.e., let's not make huge paper anymore after all.

File: render50 (hunk 5 above)
Comment (coltonoscopy) [reply render50]: Hardcoded filenames are for the temp files generated before concatenation, not the input files :p

Comment (dmalan) [reply render50]: Still messy, breaks the abstraction that is that method; everything should be cleanly encapsulated. Plus, if the student has files by the same name in their dir, we just clobbered them. Should instead use dynamically generated tempfiles in the system temp dir.

Comment (coltonoscopy) [reply render50]: Fair, though this is presumably fixed anyway by your suggestion to use `tempfile.mkstemp` above!

**Pull request closed without merging**